File: lexer.js
/**
 * LexiScan — Lexical Analyzer Core
 * =========================================
 * Implements a Deterministic Finite Automaton (DFA)-based lexer
 * that tokenizes source code into categories:
 *   a. Reserved Words
 *   b. Symbols & Punctuation
 *   c. Variables / Identifiers
 *   d. Mathematical Expressions / Operators
 *   + Bonus: Literals (numbers, strings, chars), Comments
 *
 * Supports: C, C++, Java, Python (configurable)
 */

'use strict';

/* ─── Token Types ───────────────────────────────────────────── */
const TOKEN_TYPE = Object.freeze({
  RESERVED:  'reserved',   // a. Reserved words / keywords
  SYMBOL:    'symbol',     // b. Symbols & punctuation
  VARIABLE:  'variable',   // c. Variable / identifier
  MATH:      'math',       // d. Math operators & expressions
  NUMBER:    'number',     // Numeric literals
  STRING:    'string',     // String / char literals
  COMMENT:   'comment',    // Comments
  UNKNOWN:   'unknown',    // Unrecognized
});

/* ─── Language Configurations ───────────────────────────────── */
const LANG_CONFIGS = {
  c: {
    name: 'C / C++',
    reserved: new Set([
      'auto','break','case','char','const','continue','default','do',
      'double','else','enum','extern','float','for','goto','if','inline',
      'int','long','register','restrict','return','short','signed','sizeof',
      'static','struct','switch','typedef','union','unsigned','void',
      'volatile','while','_Bool','_Complex','_Imaginary',
      // C++ additions
      'and','and_eq','asm','bitand','bitor','bool','catch','class',
      'compl','const_cast','delete','dynamic_cast','explicit','export',
      'false','friend','mutable','namespace','new','not','not_eq','nullptr',
      'operator','or','or_eq','private','protected','public','reinterpret_cast',
      'static_assert','static_cast','template','this','throw','true','try',
      'typeid','typename','using','virtual','wchar_t','xor','xor_eq',
      // Common macros/functions treated as keywords
      'printf','scanf','malloc','free','NULL','main','#include','#define',
      '#ifdef','#ifndef','#endif','#if','#else','#elif','#pragma',
    ]),
    mathOps: new Set(['+','-','*','/','%','**','=','==','!=','<','>','<=','>=',
                      '&&','||','!','&','|','^','~','<<','>>','++','--',
                      '+=','-=','*=','/=','%=','&=','|=','^=','<<=','>>=']),
    symbols: new Set(['(',')','{','}','[',']',';',',','.',':','?','#','\\',"'"]),
    stringDelimiters: ['"', "'"],
    lineComment: '//',
    blockComment: { open: '/*', close: '*/' },
  },
  python: {
    name: 'Python',
    reserved: new Set([
      'False','None','True','and','as','assert','async','await',
      'break','class','continue','def','del','elif','else','except',
      'finally','for','from','global','if','import','in','is','lambda',
      'nonlocal','not','or','pass','raise','return','try','while','with',
      'yield','print','input','range','len','type','int','float','str',
      'list','dict','tuple','set','bool','open','super','self','__init__',
      '__main__','__name__','append','extend','pop','remove','format',
    ]),
    mathOps: new Set(['+','-','*','/','//','%','**','=','==','!=','<','>','<=','>=',
                      'and','or','not','&','|','^','~','<<','>>',
                      '+=','-=','*=','/=','//=','%=','**=','&=','|=','^=','<<=','>>=']),
    symbols: new Set(['(',')','{','}','[',']',':',',','.','#','@',"'"]),
    stringDelimiters: ['"', "'", '"""', "'''"],
    lineComment: '#',
    blockComment: null,
  },
  java: {
    name: 'Java',
    reserved: new Set([
      'abstract','assert','boolean','break','byte','case','catch','char',
      'class','const','continue','default','do','double','else','enum',
      'extends','final','finally','float','for','goto','if','implements',
      'import','instanceof','int','interface','long','native','new',
      'package','private','protected','public','return','short','static',
      'strictfp','super','switch','synchronized','this','throw','throws',
      'transient','try','void','volatile','while','true','false','null',
      'String','System','out','println','print','main','args','Scanner',
      'ArrayList','HashMap','List','Map','Override','SuppressWarnings',
    ]),
    mathOps: new Set(['+','-','*','/','%','=','==','!=','<','>','<=','>=',
                      '&&','||','!','&','|','^','~','<<','>>','>>>',
                      '++','--','+=','-=','*=','/=','%=','&=','|=','^=']),
    symbols: new Set(['(',')','{','}','[',']',';',',','.','?',':','@',"'"]),
    stringDelimiters: ['"', "'"],
    lineComment: '//',
    blockComment: { open: '/*', close: '*/' },
  },
  custom: {
    name: 'Kustom (Generik)',
    reserved: new Set([
      // merge of common keywords across languages
      'if','else','while','for','do','switch','case','break','continue',
      'return','true','false','null','void','int','float','double','char',
      'string','bool','class','function','def','var','let','const',
      'new','delete','import','export','try','catch','throw','and','or','not',
      'in','is','as','from','with','pass','None','True','False',
    ]),
    mathOps: new Set(['+','-','*','/','%','**','//','=','==','!=','<','>','<=','>=',
                      '&&','||','!','&','|','^','~','<<','>>',
                      '++','--','+=','-=','*=','/=','%=']),
    symbols: new Set(['(',')','{','}','[',']',';',',','.','?',':','#','@',"'"]),
    stringDelimiters: ['"', "'"],
    lineComment: '//',
    blockComment: { open: '/*', close: '*/' },
  },
};

/* ─── DFA States ────────────────────────────────────────────── */
const STATE = Object.freeze({
  START:          0,
  IN_IDENTIFIER:  1,
  IN_NUMBER:      2,
  IN_FLOAT:       3,
  IN_STRING_DQ:   4,  // double-quote string
  IN_STRING_SQ:   5,  // single-quote string
  IN_COMMENT_LINE:6,
  IN_COMMENT_BLCK:7,
  IN_OPERATOR:    8,
  IN_PREPROC:     9,  // preprocessor (#...)
  DONE:           10,
});

/* ─── Helper predicates (character classes) ─────────────────── */
const isAlpha     = c => /^[a-zA-Z_$]$/.test(c);
const isDigit     = c => /^[0-9]$/.test(c);
const isAlphaNum  = c => /^[a-zA-Z0-9_$]$/.test(c);
const isHexDigit  = c => /^[0-9a-fA-F]$/.test(c);
const isWhitespace= c => /^[ \t\r\n]$/.test(c);
const isOperChar  = c => /^[+\-*/%=!<>&|^~]$/.test(c);

/**
 * Core Lexer class.
 * Uses a hand-written DFA to scan source code character by character.
 */
class Lexer {
  /**
   * @param {string} source  - The source code to tokenize
   * @param {string} langKey - Language key (c|python|java|custom)
   */
  constructor(source, langKey = 'c') {
    this.source  = source;
    this.lang    = LANG_CONFIGS[langKey] || LANG_CONFIGS.c;
    this.langKey = langKey;
    this.pos     = 0;
    this.line    = 1;
    this.col     = 1;
    this.tokens  = [];
    this.errors  = [];
  }

  /* ── Character access helpers ─────────────────────────────── */
  peek(offset = 0) {
    return this.source[this.pos + offset] ?? '\0';
  }

  advance() {
    const ch = this.source[this.pos++];
    if (ch === '\n') { this.line++; this.col = 1; }
    else { this.col++; }
    return ch;
  }

  /* ── Emit a token ─────────────────────────────────────────── */
  emit(value, type, startLine) {
    this.tokens.push({ value, type, line: startLine, col: this.col });
  }

  /* ── Classify a finalized token ───────────────────────────── */
  classify(raw) {
    const { reserved, mathOps, symbols } = this.lang;
    if (reserved.has(raw))  return TOKEN_TYPE.RESERVED;
    if (symbols.has(raw))   return TOKEN_TYPE.SYMBOL;
    if (mathOps.has(raw))   return TOKEN_TYPE.MATH;
    return TOKEN_TYPE.VARIABLE;
  }

  /**
   * Main tokenization method.
   * Implements a DFA traversal over the source string.
   * @returns {{ tokens: Token[], errors: string[] }}
   */
  tokenize() {
    this.tokens = [];
    this.errors = [];
    const src   = this.source;
    const n     = src.length;
    let   state = STATE.START;
    let   buf   = '';
    let   startLine = 1;

    const flush = () => {
      if (!buf) return;
      const type = this.classify(buf);
      this.emit(buf, type, startLine);
      buf = '';
    };

    while (this.pos <= n) {
      const ch  = this.pos < n ? src[this.pos] : '\0';  // current char
      const ch2 = this.pos + 1 < n ? src[this.pos + 1] : '\0';  // lookahead

      switch (state) {
        /* ── START ──────────────────────────────────────────── */
        case STATE.START: {
          if (ch === '\0') break;

          // Whitespace: skip
          if (isWhitespace(ch)) { this.advance(); break; }

          startLine = this.line;

          // Line comment
          if (this.lang.lineComment && src.startsWith(this.lang.lineComment, this.pos)) {
            this.advance(); if (this.lang.lineComment.length > 1) this.advance();
            buf = this.lang.lineComment;
            state = STATE.IN_COMMENT_LINE;
            break;
          }

          // Block comment
          if (this.lang.blockComment && src.startsWith(this.lang.blockComment.open, this.pos)) {
            buf = this.lang.blockComment.open;
            this.advance(); this.advance();
            state = STATE.IN_COMMENT_BLCK;
            break;
          }

          // Preprocessor directive (#...)
          if (ch === '#' && this.langKey === 'c') {
            buf = '#';
            this.advance();
            state = STATE.IN_PREPROC;
            break;
          }

          // String literal (double-quote)
          if (ch === '"') {
            buf = '"';
            this.advance();
            state = STATE.IN_STRING_DQ;
            break;
          }

          // String/char literal (single-quote)
          if (ch === "'") {
            buf = "'";
            this.advance();
            state = STATE.IN_STRING_SQ;
            break;
          }

          // Identifier or keyword
          if (isAlpha(ch)) {
            buf = ch;
            this.advance();
            state = STATE.IN_IDENTIFIER;
            break;
          }

          // Hex literal (0x...)
          if (ch === '0' && (ch2 === 'x' || ch2 === 'X')) {
            buf = '0' + ch2;
            this.advance(); this.advance();
            while (this.pos < n && isHexDigit(src[this.pos])) buf += this.advance();
            this.emit(buf, TOKEN_TYPE.NUMBER, startLine);
            buf = '';
            break;
          }

          // Numeric literal
          if (isDigit(ch) || (ch === '.' && isDigit(ch2))) {
            buf = ch;
            this.advance();
            state = STATE.IN_NUMBER;
            break;
          }

          // Negative number after operator or start
          if (ch === '-' && isDigit(ch2) && this.tokens.length === 0) {
            buf = ch;
            this.advance();
            state = STATE.IN_NUMBER;
            break;
          }

          // Multi-char operators (try longest match first)
          const multiOps = ['>>>=','<<=','>>=','...','**=','//=',
                             '===','!==','>>>','<<=',
                             '**','//','==','!=','<=','>=','&&','||',
                             '++','--','+=','-=','*=','/=','%=','&=','|=','^=','<<','>>','->'
                            ];
          let matched = false;
          for (const op of multiOps) {
            if (src.startsWith(op, this.pos) && this.lang.mathOps.has(op)) {
              this.emit(op, TOKEN_TYPE.MATH, startLine);
              for (let i = 0; i < op.length; i++) this.advance();
              matched = true;
              break;
            }
          }
          if (matched) break;

          // Single-char operator
          if (isOperChar(ch)) {
            const op1 = ch;
            this.advance();
            if (this.lang.mathOps.has(op1)) {
              this.emit(op1, TOKEN_TYPE.MATH, startLine);
            } else {
              this.emit(op1, TOKEN_TYPE.SYMBOL, startLine);
            }
            break;
          }

          // Symbol / punctuation
          if (this.lang.symbols.has(ch) || /^[(){}[\];,.:?@\\]$/.test(ch)) {
            this.emit(ch, TOKEN_TYPE.SYMBOL, startLine);
            this.advance();
            break;
          }

          // Unknown character — emit as unknown but don't stop
          this.emit(ch, TOKEN_TYPE.UNKNOWN, startLine);
          this.advance();
          break;
        }

        /* ── IN_IDENTIFIER ──────────────────────────────────── */
        case STATE.IN_IDENTIFIER: {
          if (this.pos < n && isAlphaNum(ch)) {
            buf += this.advance();
          } else {
            flush();
            state = STATE.START;
          }
          break;
        }

        /* ── IN_NUMBER ──────────────────────────────────────── */
        case STATE.IN_NUMBER: {
          if (this.pos < n && isDigit(ch)) {
            buf += this.advance();
          } else if (ch === '.' && ch2 !== '.') {
            buf += this.advance();
            state = STATE.IN_FLOAT;
          } else if ((ch === 'e' || ch === 'E') && (isDigit(ch2) || ch2 === '+' || ch2 === '-')) {
            buf += this.advance();
            if (src[this.pos] === '+' || src[this.pos] === '-') buf += this.advance();
            state = STATE.IN_FLOAT;
          } else if ((ch === 'f' || ch === 'F' || ch === 'l' || ch === 'L' ||
                      ch === 'u' || ch === 'U') && this.pos < n) {
            buf += this.advance();
            this.emit(buf, TOKEN_TYPE.NUMBER, startLine);
            buf = ''; state = STATE.START;
          } else {
            this.emit(buf, TOKEN_TYPE.NUMBER, startLine);
            buf = ''; state = STATE.START;
          }
          break;
        }

        /* ── IN_FLOAT ───────────────────────────────────────── */
        case STATE.IN_FLOAT: {
          if (this.pos < n && isDigit(ch)) {
            buf += this.advance();
          } else if ((ch === 'f' || ch === 'F' || ch === 'l' || ch === 'L') && this.pos < n) {
            buf += this.advance();
            this.emit(buf, TOKEN_TYPE.NUMBER, startLine);
            buf = ''; state = STATE.START;
          } else {
            this.emit(buf, TOKEN_TYPE.NUMBER, startLine);
            buf = ''; state = STATE.START;
          }
          break;
        }

        /* ── IN_STRING_DQ ───────────────────────────────────── */
        case STATE.IN_STRING_DQ: {
          if (this.pos >= n) {
            this.errors.push(`Line ${startLine}: String literal tidak ditutup`);
            this.emit(buf, TOKEN_TYPE.STRING, startLine);
            buf = ''; state = STATE.START;
          } else if (ch === '\\') {
            buf += this.advance() + this.advance(); // escape sequence
          } else if (ch === '"') {
            buf += this.advance();
            this.emit(buf, TOKEN_TYPE.STRING, startLine);
            buf = ''; state = STATE.START;
          } else {
            buf += this.advance();
          }
          break;
        }

        /* ── IN_STRING_SQ ───────────────────────────────────── */
        case STATE.IN_STRING_SQ: {
          if (this.pos >= n) {
            this.errors.push(`Line ${startLine}: Char/string literal tidak ditutup`);
            this.emit(buf, TOKEN_TYPE.STRING, startLine);
            buf = ''; state = STATE.START;
          } else if (ch === '\\') {
            buf += this.advance() + this.advance();
          } else if (ch === "'") {
            buf += this.advance();
            this.emit(buf, TOKEN_TYPE.STRING, startLine);
            buf = ''; state = STATE.START;
          } else {
            buf += this.advance();
          }
          break;
        }

        /* ── IN_COMMENT_LINE ────────────────────────────────── */
        case STATE.IN_COMMENT_LINE: {
          if (this.pos >= n || ch === '\n') {
            this.emit(buf, TOKEN_TYPE.COMMENT, startLine);
            buf = ''; state = STATE.START;
          } else {
            buf += this.advance();
          }
          break;
        }

        /* ── IN_COMMENT_BLCK ────────────────────────────────── */
        case STATE.IN_COMMENT_BLCK: {
          if (this.pos >= n) {
            this.errors.push(`Line ${startLine}: Komentar blok tidak ditutup`);
            this.emit(buf, TOKEN_TYPE.COMMENT, startLine);
            buf = ''; state = STATE.START;
          } else if (src.startsWith(this.lang.blockComment.close, this.pos)) {
            buf += this.lang.blockComment.close;
            this.advance(); this.advance();
            this.emit(buf, TOKEN_TYPE.COMMENT, startLine);
            buf = ''; state = STATE.START;
          } else {
            buf += this.advance();
          }
          break;
        }

        /* ── IN_PREPROC ─────────────────────────────────────── */
        case STATE.IN_PREPROC: {
          if (this.pos >= n || ch === '\n') {
            // Check if it's a known directive
            const directive = buf.trim().split(/\s+/)[0];
            const type = this.lang.reserved.has(directive) ? TOKEN_TYPE.RESERVED : TOKEN_TYPE.RESERVED;
            this.emit(buf.trim(), type, startLine);
            buf = ''; state = STATE.START;
          } else {
            buf += this.advance();
          }
          break;
        }

        default:
          this.pos++;
          break;
      }

      // EOF
      if (this.pos >= n && state === STATE.START) break;
      if (this.pos >= n && buf) {
        // Flush remaining buffer
        if (state === STATE.IN_IDENTIFIER || state === STATE.IN_PREPROC) flush();
        else if (state === STATE.IN_NUMBER || state === STATE.IN_FLOAT)
          this.emit(buf, TOKEN_TYPE.NUMBER, startLine);
        else if (state === STATE.IN_STRING_DQ || state === STATE.IN_STRING_SQ)
          this.emit(buf, TOKEN_TYPE.STRING, startLine);
        else if (state === STATE.IN_COMMENT_LINE || state === STATE.IN_COMMENT_BLCK)
          this.emit(buf, TOKEN_TYPE.COMMENT, startLine);
        buf = '';
        break;
      }
    }

    return { tokens: this.tokens, errors: this.errors };
  }
}

/**
 * Post-process tokens:
 * - Identify mathematical expressions (sequences with operators)
 * - De-duplicate for unique token list
 * - Build grouped output
 * @param {Array} tokens
 * @returns {Object} Grouped results
 */
function groupTokens(tokens) {
  const groups = {
    [TOKEN_TYPE.RESERVED]: [],
    [TOKEN_TYPE.SYMBOL]:   [],
    [TOKEN_TYPE.VARIABLE]: [],
    [TOKEN_TYPE.MATH]:     [],
    [TOKEN_TYPE.NUMBER]:   [],
    [TOKEN_TYPE.STRING]:   [],
    [TOKEN_TYPE.COMMENT]:  [],
    [TOKEN_TYPE.UNKNOWN]:  [],
  };

  // Identify math expressions: sequences of identifiers/numbers + math ops
  // Mark tokens that are part of a math context
  const mathExprTokens = new Set();
  for (let i = 0; i < tokens.length; i++) {
    if (tokens[i].type === TOKEN_TYPE.MATH) {
      // Mark neighboring identifiers/numbers as math-context
      if (i > 0 && (tokens[i-1].type === TOKEN_TYPE.VARIABLE || tokens[i-1].type === TOKEN_TYPE.NUMBER)) {
        mathExprTokens.add(i - 1);
      }
      if (i < tokens.length - 1 && (tokens[i+1].type === TOKEN_TYPE.VARIABLE || tokens[i+1].type === TOKEN_TYPE.NUMBER)) {
        mathExprTokens.add(i + 1);
      }
      mathExprTokens.add(i);
    }
  }

  for (let i = 0; i < tokens.length; i++) {
    const t = tokens[i];
    let effectiveType = t.type;

    // If a variable/number is adjacent to a math op, promote to math group too
    if (mathExprTokens.has(i) && effectiveType !== TOKEN_TYPE.MATH) {
      effectiveType = TOKEN_TYPE.MATH;
    }

    if (!groups[effectiveType]) groups[effectiveType] = [];
    groups[effectiveType].push({ ...t, effectiveType });
  }

  return groups;
}

/**
 * Get unique tokens (value+type) with counts.
 */
function uniqueTokens(tokenList) {
  const map = new Map();
  for (const t of tokenList) {
    const key = t.value + '|' + t.type;
    if (map.has(key)) {
      map.get(key).count++;
    } else {
      map.set(key, { ...t, count: 1 });
    }
  }
  return [...map.values()];
}

/* ─── Export ──────────────────────────────────────────────── */
window.Lexer        = Lexer;
window.groupTokens  = groupTokens;
window.uniqueTokens = uniqueTokens;
window.TOKEN_TYPE   = TOKEN_TYPE;
window.LANG_CONFIGS = LANG_CONFIGS;

File: core/lexer/lexer.go
package lexer

import (
	"strings"
	"unicode"
)

// ── DFA States ────────────────────────────────────────────────
type state int

const (
	stStart       state = iota
	stIdentifier        // reading identifier or keyword
	stNumber            // reading integer part
	stFloat             // reading float fraction / exponent
	stStringDQ          // inside double-quoted string
	stStringSQ          // inside single-quoted string
	stCommentLine       // inside // or # line comment
	stCommentBlk        // inside /* block comment */
	stPreproc           // inside #directive (C only)
)

// ── Lexer ─────────────────────────────────────────────────────

// Lexer tokenizes source code using a DFA.
type Lexer struct {
	src  []rune
	pos  int
	line int
	lang *LangConfig

	tokens []Token
	errors []string
}

// NewLexer creates a Lexer for the given source and language key.
func NewLexer(src, langKey string) *Lexer {
	return &Lexer{
		src:  []rune(src),
		pos:  0,
		line: 1,
		lang: GetLang(langKey),
	}
}

// ── helpers ───────────────────────────────────────────────────

func (l *Lexer) peek(offset int) rune {
	i := l.pos + offset
	if i >= len(l.src) {
		return 0
	}
	return l.src[i]
}

func (l *Lexer) advance() rune {
	ch := l.src[l.pos]
	l.pos++
	if ch == '\n' {
		l.line++
	}
	return ch
}

func (l *Lexer) startsWith(s string) bool {
	runes := []rune(s)
	if l.pos+len(runes) > len(l.src) {
		return false
	}
	for i, r := range runes {
		if l.src[l.pos+i] != r {
			return false
		}
	}
	return true
}

func (l *Lexer) skipRunes(n int) {
	for i := 0; i < n; i++ {
		l.advance()
	}
}

func (l *Lexer) emit(value string, typ TokenType, line int) {
	if strings.TrimSpace(value) == "" {
		return
	}
	l.tokens = append(l.tokens, Token{Value: value, Type: typ, Line: line})
}

// classify decides the final type of an identifier-like token.
func (l *Lexer) classify(raw string) TokenType {
	if l.lang.Reserved[raw] {
		return RESERVED
	}
	if l.lang.MathOps[raw] {
		return MATH
	}
	if l.lang.Symbols[raw] {
		return SYMBOL
	}
	return VARIABLE
}

// ── Main DFA ──────────────────────────────────────────────────

// Tokenize runs the DFA and returns all tokens and any errors.
func (l *Lexer) Tokenize() ([]Token, []string) {
	src := l.src
	n := len(src)

	var (
		cur       state = stStart
		buf       []rune
		startLine int
	)

	flush := func() {
		if len(buf) == 0 {
			return
		}
		raw := string(buf)
		l.emit(raw, l.classify(raw), startLine)
		buf = buf[:0]
	}

	for l.pos <= n {
		ch := rune(0)
		if l.pos < n {
			ch = src[l.pos]
		}

		switch cur {

		// ── START ────────────────────────────────────────────
		case stStart:
			if l.pos >= n {
				goto done
			}

			// Whitespace
			if unicode.IsSpace(ch) {
				l.advance()
				continue
			}

			startLine = l.line

			// Block comment
			if l.lang.BlockOpen != "" && l.startsWith(l.lang.BlockOpen) {
				buf = append(buf, []rune(l.lang.BlockOpen)...)
				l.skipRunes(len([]rune(l.lang.BlockOpen)))
				cur = stCommentBlk
				continue
			}

			// Line comment
			if l.lang.LineComment != "" && l.startsWith(l.lang.LineComment) {
				buf = append(buf, []rune(l.lang.LineComment)...)
				l.skipRunes(len([]rune(l.lang.LineComment)))
				cur = stCommentLine
				continue
			}

			// Preprocessor directive (#include, etc.) — C only
			if l.lang.IsPreproc && ch == '#' {
				buf = append(buf, l.advance())
				cur = stPreproc
				continue
			}

			// String: double-quote
			if ch == '"' {
				buf = append(buf, l.advance())
				cur = stStringDQ
				continue
			}

			// String/char: single-quote
			if ch == '\'' {
				buf = append(buf, l.advance())
				cur = stStringSQ
				continue
			}

			// Identifier / keyword
			if unicode.IsLetter(ch) || ch == '_' || ch == '$' {
				buf = append(buf, l.advance())
				cur = stIdentifier
				continue
			}

			// Hex literal 0x...
			if ch == '0' && l.pos+1 < n && (src[l.pos+1] == 'x' || src[l.pos+1] == 'X') {
				buf = append(buf, l.advance(), l.advance()) // 0 x
				for l.pos < n && isHex(src[l.pos]) {
					buf = append(buf, l.advance())
				}
				l.emit(string(buf), NUMBER, startLine)
				buf = buf[:0]
				continue
			}

			// Numeric literal
			if unicode.IsDigit(ch) || (ch == '.' && l.pos+1 < n && unicode.IsDigit(src[l.pos+1])) {
				buf = append(buf, l.advance())
				cur = stNumber
				continue
			}

			// Multi-char operators (longest match first)
			if matched, op := l.matchMultiOp(); matched {
				if l.lang.MathOps[op] {
					l.emit(op, MATH, startLine)
				} else {
					l.emit(op, SYMBOL, startLine)
				}
				continue
			}

			// Single-char operator
			if isOperChar(ch) {
				op := string([]rune{l.advance()})
				if l.lang.MathOps[op] {
					l.emit(op, MATH, startLine)
				} else {
					l.emit(op, SYMBOL, startLine)
				}
				continue
			}

			// Symbol / punctuation
			sym := string([]rune{ch})
			if l.lang.Symbols[sym] || isPunct(ch) {
				l.emit(sym, SYMBOL, startLine)
				l.advance()
				continue
			}

			// Unknown character
			l.emit(sym, UNKNOWN, startLine)
			l.advance()

		// ── IDENTIFIER ───────────────────────────────────────
		case stIdentifier:
			if l.pos < n && (unicode.IsLetter(ch) || unicode.IsDigit(ch) || ch == '_' || ch == '$') {
				buf = append(buf, l.advance())
			} else {
				flush()
				cur = stStart
			}

		// ── NUMBER (integer part) ────────────────────────────
		case stNumber:
			if l.pos < n && unicode.IsDigit(ch) {
				buf = append(buf, l.advance())
			} else if ch == '.' && l.pos+1 < n && ch != '.' {
				buf = append(buf, l.advance())
				cur = stFloat
			} else if (ch == 'e' || ch == 'E') && l.pos+1 < n {
				buf = append(buf, l.advance())
				if l.pos < n && (src[l.pos] == '+' || src[l.pos] == '-') {
					buf = append(buf, l.advance())
				}
				cur = stFloat
			} else if l.pos < n && (ch == 'f' || ch == 'F' || ch == 'l' || ch == 'L' || ch == 'u' || ch == 'U') {
				buf = append(buf, l.advance())
				l.emit(string(buf), NUMBER, startLine)
				buf = buf[:0]
				cur = stStart
			} else {
				l.emit(string(buf), NUMBER, startLine)
				buf = buf[:0]
				cur = stStart
			}

		// ── FLOAT ────────────────────────────────────────────
		case stFloat:
			if l.pos < n && unicode.IsDigit(ch) {
				buf = append(buf, l.advance())
			} else if l.pos < n && (ch == 'f' || ch == 'F' || ch == 'l' || ch == 'L') {
				buf = append(buf, l.advance())
				l.emit(string(buf), NUMBER, startLine)
				buf = buf[:0]
				cur = stStart
			} else {
				l.emit(string(buf), NUMBER, startLine)
				buf = buf[:0]
				cur = stStart
			}

		// ── STRING (double-quote) ─────────────────────────────
		case stStringDQ:
			if l.pos >= n {
				l.errors = append(l.errors, formatErr(startLine, "String literal tidak ditutup"))
				l.emit(string(buf), STRING, startLine)
				buf = buf[:0]
				cur = stStart
			} else if ch == '\\' && l.pos+1 < n {
				buf = append(buf, l.advance(), l.advance()) // escape sequence
			} else if ch == '"' {
				buf = append(buf, l.advance())
				l.emit(string(buf), STRING, startLine)
				buf = buf[:0]
				cur = stStart
			} else {
				buf = append(buf, l.advance())
			}

		// ── STRING (single-quote) ─────────────────────────────
		case stStringSQ:
			if l.pos >= n {
				l.errors = append(l.errors, formatErr(startLine, "Char/string literal tidak ditutup"))
				l.emit(string(buf), STRING, startLine)
				buf = buf[:0]
				cur = stStart
			} else if ch == '\\' && l.pos+1 < n {
				buf = append(buf, l.advance(), l.advance())
			} else if ch == '\'' {
				buf = append(buf, l.advance())
				l.emit(string(buf), STRING, startLine)
				buf = buf[:0]
				cur = stStart
			} else {
				buf = append(buf, l.advance())
			}

		// ── LINE COMMENT ──────────────────────────────────────
		case stCommentLine:
			if l.pos >= n || ch == '\n' {
				l.emit(string(buf), COMMENT, startLine)
				buf = buf[:0]
				cur = stStart
			} else {
				buf = append(buf, l.advance())
			}

		// ── BLOCK COMMENT ─────────────────────────────────────
		case stCommentBlk:
			if l.pos >= n {
				l.errors = append(l.errors, formatErr(startLine, "Block comment tidak ditutup"))
				l.emit(string(buf), COMMENT, startLine)
				buf = buf[:0]
				cur = stStart
			} else if l.startsWith(l.lang.BlockClose) {
				buf = append(buf, []rune(l.lang.BlockClose)...)
				l.skipRunes(len([]rune(l.lang.BlockClose)))
				l.emit(string(buf), COMMENT, startLine)
				buf = buf[:0]
				cur = stStart
			} else {
				buf = append(buf, l.advance())
			}

		// ── PREPROCESSOR (#include …) ─────────────────────────
		case stPreproc:
			if l.pos >= n || ch == '\n' {
				raw := strings.TrimSpace(string(buf))
				l.emit(raw, RESERVED, startLine)
				buf = buf[:0]
				cur = stStart
			} else {
				buf = append(buf, l.advance())
			}
		}
	}

done:
	// Flush any remaining buffer
	if len(buf) > 0 {
		switch cur {
		case stIdentifier, stPreproc:
			flush()
		case stNumber, stFloat:
			l.emit(string(buf), NUMBER, startLine)
		case stStringDQ, stStringSQ:
			l.emit(string(buf), STRING, startLine)
		case stCommentLine, stCommentBlk:
			l.emit(string(buf), COMMENT, startLine)
		}
	}

	return l.tokens, l.errors
}

// matchMultiOp tries to match a multi-character operator at the current pos.
func (l *Lexer) matchMultiOp() (bool, string) {
	candidates := []string{
		">>>=", "**=", "//=", "<<=", ">>=",
		"===", "!==", ">>>",
		"**", "//", "==", "!=", "<=", ">=",
		"&&", "||", "++", "--",
		"+=", "-=", "*=", "/=", "%=", "&=", "|=", "^=",
		"<<", ">>", "->", "...",
	}
	for _, op := range candidates {
		if l.startsWith(op) {
			l.skipRunes(len([]rune(op)))
			return true, op
		}
	}
	return false, ""
}

// ── Grouping ──────────────────────────────────────────────────

// GroupTokens groups tokens by type and marks math-adjacent tokens.
func GroupTokens(tokens []Token) map[string][]Token {
	groups := map[string][]Token{
		string(RESERVED): {},
		string(SYMBOL):   {},
		string(VARIABLE): {},
		string(MATH):     {},
		string(NUMBER):   {},
		string(STRING):   {},
		string(COMMENT):  {},
		string(UNKNOWN):  {},
	}

	// Mark indices adjacent to MATH operators
	mathAdj := make(map[int]bool)
	for i, t := range tokens {
		if t.Type == MATH {
			if i > 0 && (tokens[i-1].Type == VARIABLE || tokens[i-1].Type == NUMBER) {
				mathAdj[i-1] = true
			}
			if i < len(tokens)-1 && (tokens[i+1].Type == VARIABLE || tokens[i+1].Type == NUMBER) {
				mathAdj[i+1] = true
			}
			mathAdj[i] = true
		}
	}

	for i, t := range tokens {
		eff := t.Type
		if mathAdj[i] && eff != MATH {
			eff = MATH
		}
		t.Type = eff
		key := string(eff)
		if _, ok := groups[key]; !ok {
			groups[key] = []Token{}
		}
		groups[key] = append(groups[key], t)
	}
	return groups
}

// BuildStats counts tokens per category.
func BuildStats(groups map[string][]Token) Stats {
	total := 0
	for _, g := range groups {
		total += len(g)
	}
	return Stats{
		Total:    total,
		Reserved: len(groups[string(RESERVED)]),
		Symbol:   len(groups[string(SYMBOL)]),
		Variable: len(groups[string(VARIABLE)]),
		Math:     len(groups[string(MATH)]),
		Number:   len(groups[string(NUMBER)]),
		String:   len(groups[string(STRING)]),
		Comment:  len(groups[string(COMMENT)]),
	}
}

// ── Character helpers ─────────────────────────────────────────

func isHex(r rune) bool {
	return (r >= '0' && r <= '9') ||
		(r >= 'a' && r <= 'f') ||
		(r >= 'A' && r <= 'F')
}

func isOperChar(r rune) bool {
	return strings.ContainsRune("+-*/%=!<>&|^~", r)
}

func isPunct(r rune) bool {
	return strings.ContainsRune("(){}[];,.:?@\\#'", r)
}

func formatErr(line int, msg string) string {
	return "Baris " + itoa(line) + ": " + msg
}

func itoa(n int) string {
	if n == 0 {
		return "0"
	}
	s := ""
	for n > 0 {
		s = string(rune('0'+n%10)) + s
		n /= 10
	}
	return s
}
